### ISSUE TITLE
K8SSAND-924 ⁃ Strange image pull behaviour on master

### ISSUE THREAD

**Issue by Miles-Garnsey** [k8ssandra/cass-operator#179]
It seems that the current `master` branch isn't allowing me to override the container `imagePullpolicy`. 1.8.x introduces a new field `spec. imageConfig. imagePullPolicy` which may not be behaving as intended.

**Problem 1: container imagePullPolicy is dropped from the cassDC CR**

So I create a cassDC with this:

```
spec:
  podTemplateSpec:
    spec:
      containers:
        - name: "cassandra"
          image: k8ssandra/mgmtapi-4_0:latest
          imagePullpolicy: IfNotPresent
```

But when it gets created the field has been dropped:

```
$kubectl get -o yaml -n cass-operator cassandradatacenters.cassandra.datastax.com dc1

spec:
  podTemplateSpec:
    spec:
      containers:
      - image: k8ssandra/mgmtapi-4_0:latest
        name: cassandra
        resources: {}
```

(EDIT: this ticket originally talked about a second issue to do with imageConfig - this was user error not a bug).

This is a blocker for #175 as it means I cannot load my locally built management API image using `kind load` due to the `imagePullPolicy` reverting to `Always` during reconciliation of the STS.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-924) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-924
┆priority: Medium

**Comment by Miles-Garnsey** [comment]
Progress so far in diagnosing this: 

1. I've inserted some logging and I'm seeing `2021-09-24T08:46:39.213Z	INFO	reconciliation_handler	cassContainer is 	{"cassContainer": "&Container{Name:cassandra,Image:k8ssandra/mgmtapi-4_0:latest,Command:[],Args:[],WorkingDir:,Ports:[]ContainerPort{},Env:[]EnvVar{},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[]VolumeMount{},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,}"}`. (at [this line](https://github.com/Miles-Garnsey/cass-operator/blob/fc9d796da537cd33351cd3fb5be21ea59079c4f1/pkg/reconciliation/construct_podtemplatespec.go#L408)).
2. This is not quite right as cassDC is set up with:
```
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
  namespace: cass-operator
spec:
  podTemplateSpec:
    spec:
      containers:
        - name: "cassandra"
          image: k8ssandra/mgmtapi-4_0:latest
          imagePullpolicy: IfNotPresent
```

I think this field is not coming through into the cassContainer object. I'll investigate further if no one else picks this up.

**Comment by Miles-Garnsey** [comment]
Ran the same test several times just now. The `imagePullpolicy: IfNotPresent` field appeared in both the cassDC and in the sts it manages. 

I did rewrite the manifest a bit while trying to diagnose. I'm guessing I must have made a typo or something in the original one which has silently failed due to there not being much validation on the CassDC CR. I know that we're thinking about how to fix this in future via an additional config CR (which would then enable the cassDC to be more rigorously validated). 

I'm going to close this for now.
